cpp/src/arrow/flight/sql/server_session_middleware.cc

```diff
@@ -0,0 +1,179 @@
+    const std::string pair_sep = "=";
+    const size_t pair_sep_len = pair_sep.length();
+
+    std::vector<std::pair<std::string, std::string>> result;
+
+    size_t cur = 0;
+    while (cur < s.length()) {
+      const size_t end = s.find(list_sep, cur);
+      size_t len;
+      if (end == std::string::npos) {
+        // No (further) list delimiters
+        len = std::string::npos;
+        cur = s.length();
+      } else {
+        len = end - cur;
+        cur = end;
+      }
+      const std::string_view tok = s.substr(cur, len);
+
+      const size_t val_pos = tok.find(pair_sep);
+      result.emplace_back(
+          tok.substr(0, val_pos),
+          tok.substr(val_pos + pair_sep_len, std::string::npos)
+      );
+    }
+
+    return result;
+  }
+
+ public:
+  Status StartCall(const CallInfo &, const CallHeaders &incoming_headers,
+                   std::shared_ptr<ServerMiddleware> *middleware) {
+    std::string session_id;
+
+    const std::pair<CallHeaders::const_iterator, CallHeaders::const_iterator>&
+        headers_it_pr = incoming_headers.equal_range("cookie");
+    for (auto itr = headers_it_pr.first; itr != headers_it_pr.second; ++itr) {
+      const std::string_view& cookie_header = itr->second;
+      const std::vector<std::pair<std::string, std::string>> cookies =
+          ParseCookieString(cookie_header);
+      for (const std::pair<std::string, std::string>& cookie : cookies) {
+        if (cookie.first == kSessionCookieName) {
+          session_id = cookie.second;
+          if (!session_id.length())
+            return Status::Invalid(
+                "Empty " + static_cast<std::string>(kSessionCookieName)
+                + " cookie value.");
+        }
+      }
+      if (session_id.length()) break;
+    }
+
+    if (!session_id.length()) {
+      // No cookie was found
+      *middleware = std::shared_ptr<ServerSessionMiddleware>(
+          new ServerSessionMiddleware(this, incoming_headers));
+    } else {
+      try {
+        const std::shared_lock<std::shared_mutex> l(session_store_lock_);
+        auto session = session_store_.at(session_id);
+        *middleware = std::shared_ptr<ServerSessionMiddleware>(
+            new ServerSessionMiddleware(this, incoming_headers,
+                                        session, session_id));
+      } catch (std::out_of_range& e) {
+        return Status::Invalid(
+            "Invalid or expired "
+            + static_cast<std::string>(kSessionCookieName) + " cookie.");
+      }
+    }
+
+    return Status::OK();
+  }
+
+  /// \brief Get a new, empty session option map and its id key.
+  std::shared_ptr<FlightSqlSession> GetNewSession(std::string* session_id) {
+    std::string new_id = boost::lexical_cast<std::string>(uuid_generator_());
+    *session_id = new_id;
+    auto session = std::make_shared<FlightSqlSession>();
+
+    const std::unique_lock<std::shared_mutex> l(session_store_lock_);
+    session_store_[new_id] = session;
```

Comment:
Can emplace be used here instead?

Reply:
Looks like you can only emplace the value_type (pair) in a map, and it's copying anyways and has to because I subsequently return 'session'.